File: backend/src/routes/rejestry.js
'use strict';
const express = require('express');
const db = require('../config/db');
const { authenticate } = require('../middleware/auth');
const { roleGuard } = require('../middleware/roleGuard');
const { sendXlsxResponse, sendCsvResponse } = require('../services/excelService');

const router = express.Router();

// 1. IT Systems register
router.get('/systemy', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, format } = req.query;
  let q = `SELECT s.id, s.nazwa, t.nazwa as tenant_nazwa, s.poziom_krytycznosci,
                  s.wlasciciel, s.aktywny, s.data_dodania,
                  COUNT(DISTINCT up.pracownik_id) FILTER (WHERE up.aktywne=TRUE) as aktywni_uzytkownicy,
                  MAX(pr.data_zakonczenia) as ostatni_przeglad
             FROM systemy_it s
             LEFT JOIN tenants t ON t.id = s.tenant_id
             LEFT JOIN uprawnienia up ON up.system_id = s.id
             LEFT JOIN przeglady pr ON pr.tenant_id = s.tenant_id AND pr.status='ZAKOŃCZONY'
            WHERE s.aktywny=TRUE`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND (s.tenant_id=$${params.length} OR s.tenant_id IS NULL)`; }
  q += ' GROUP BY s.id, t.nazwa ORDER BY s.poziom_krytycznosci DESC, s.nazwa';
  const { rows } = await db.query(q, params);

  const cols = [
    { header: 'System', key: 'nazwa', width: 25 },
    { header: 'Jednostka', key: 'tenant_nazwa', width: 20 },
    { header: 'Krytyczność', key: 'poziom_krytycznosci', width: 14 },
    { header: 'Właściciel', key: 'wlasciciel', width: 20 },
    { header: 'Aktywni użytkownicy', key: 'aktywni_uzytkownicy', width: 20 },
    { header: 'Ostatni przegląd', key: 'ostatni_przeglad', width: 18 },
    { header: 'Status', key: 'aktywny', width: 10 },
  ];
  if (format === 'xlsx') return sendXlsxResponse(res, cols, rows, 'Rejestr_systemow', 'Systemy IT');
  if (format === 'csv') return sendCsvResponse(res, cols, rows, 'Rejestr_systemow');
  res.json(rows);
});

// 2. Privileged users register
router.get('/uprzywilejowani', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, format } = req.query;
  let q = `SELECT up.id, up.data_od, up.aktywne,
                  u.imie || ' ' || u.nazwisko as pracownik_nazwa,
                  t.nazwa as tenant_nazwa,
                  s.nazwa as system_nazwa,
                  z.nazwa as zakres_nazwa,
                  w.numer
             FROM uprawnienia up
             JOIN pracownicy p ON p.id = up.pracownik_id
             JOIN uzytkownicy u ON u.id = p.uzytkownik_id
             JOIN tenants t ON t.id = up.tenant_id
             JOIN systemy_it s ON s.id = up.system_id
             JOIN zakres_uprawnien z ON z.id = up.zakres_id
             LEFT JOIN wnioski w ON w.id = up.wniosek_id
            WHERE z.uprzywilejowany=TRUE AND up.aktywne=TRUE`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND up.tenant_id=$${params.length}`; }
  q += ' ORDER BY t.nazwa, u.nazwisko';
  const { rows } = await db.query(q, params);

  const cols = [
    { header: 'Pracownik', key: 'pracownik_nazwa', width: 25 },
    { header: 'Jednostka', key: 'tenant_nazwa', width: 20 },
    { header: 'System', key: 'system_nazwa', width: 20 },
    { header: 'Zakres', key: 'zakres_nazwa', width: 18 },
    { header: 'Data nadania', key: 'data_od', width: 14 },
    { header: 'Nr wniosku', key: 'numer', width: 16 },
    { header: 'Aktywne', key: 'aktywne', width: 10 },
  ];
  if (format === 'xlsx') return sendXlsxResponse(res, cols, rows, 'Uprzywilejowani', 'Uprzywilejowani');
  if (format === 'csv') return sendCsvResponse(res, cols, rows, 'Uprzywilejowani');
  res.json(rows);
});

// 3. Reviews register
router.get('/przeglady', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, format } = req.query;
  let q = `SELECT pr.id, pr.typ, pr.status, pr.data_rozpoczecia, pr.data_zakonczenia,
                  t.nazwa as tenant_nazwa,
                  u.imie || ' ' || u.nazwisko as inicjujacy_nazwa,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='POZOSTAW') as pozostaw,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='COFNIJ') as cofnij,
                  COUNT(pp.id) FILTER (WHERE pp.decyzja='ZMODYFIKUJ') as zmodyfikuj
             FROM przeglady pr
             JOIN tenants t ON t.id = pr.tenant_id
             JOIN uzytkownicy u ON u.id = pr.inicjujacy_id
             LEFT JOIN pozycje_przegladu pp ON pp.przeglad_id = pr.id
            WHERE 1=1`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND pr.tenant_id=$${params.length}`; }
  q += ' GROUP BY pr.id, t.nazwa, u.imie, u.nazwisko ORDER BY pr.data_rozpoczecia DESC';
  const { rows } = await db.query(q, params);

  const cols = [
    { header: 'Nr', key: 'id', width: 8 },
    { header: 'Jednostka', key: 'tenant_nazwa', width: 20 },
    { header: 'Typ', key: 'typ', width: 12 },
    { header: 'Status', key: 'status', width: 12 },
    { header: 'Data rozp.', key: 'data_rozpoczecia', width: 14 },
    { header: 'Data zak.', key: 'data_zakonczenia', width: 14 },
    { header: 'Inicjujący', key: 'inicjujacy_nazwa', width: 22 },
    { header: 'Pozostaw', key: 'pozostaw', width: 12 },
    { header: 'Cofnij', key: 'cofnij', width: 10 },
    { header: 'Zmodyfikuj', key: 'zmodyfikuj', width: 12 },
  ];
  if (format === 'xlsx') return sendXlsxResponse(res, cols, rows, 'Rejestr_przeglady', 'Przeglądy');
  if (format === 'csv') return sendCsvResponse(res, cols, rows, 'Rejestr_przeglady');
  res.json(rows);
});

// 4. Audit log
router.get('/audit', authenticate, roleGuard('IT_ADMIN', 'SUPERADMIN'), async (req, res) => {
  const { tenant_id, user_id, akcja, tabela, from, to, page = 1, limit = 50, format } = req.query;
  const offset = (parseInt(page) - 1) * parseInt(limit);

  let q = `SELECT al.id, al.timestamp, al.akcja, al.tabela_docelowa, al.rekord_id,
                  al.username, al.rola, al.ip_adres,
                  t.nazwa as tenant_nazwa,
                  al.stare_dane, al.nowe_dane
             FROM audit_log al
             LEFT JOIN tenants t ON t.id = al.tenant_id
            WHERE 1=1`;
  const params = [];
  if (tenant_id) { params.push(tenant_id); q += ` AND al.tenant_id=$${params.length}`; }
  if (user_id) { params.push(user_id); q += ` AND al.user_id=$${params.length}`; }
  if (akcja) { params.push(`%${akcja}%`); q += ` AND al.akcja ILIKE $${params.length}`; }
  if (tabela) { params.push(tabela); q += ` AND al.tabela_docelowa=$${params.length}`; }
  if (from) { params.push(from); q += ` AND al.timestamp>=$${params.length}`; }
  if (to) { params.push(to); q += ` AND al.timestamp<=$${params.length}`; }
  q += ' ORDER BY al.timestamp DESC';

  if (!format) {
    const cq = `SELECT COUNT(*) as cnt FROM (${q}) sub`;
    const { rows: cRows } = await db.query(cq, params);
    params.push(parseInt(limit)); q += ` LIMIT $${params.length}`;
    params.push(offset); q += ` OFFSET $${params.length}`;
    const { rows } = await db.query(q, params);
    return res.json({ data: rows, total: parseInt(cRows[0].cnt), page: parseInt(page), limit: parseInt(limit) });
  }

  const { rows } = await db.query(q, params);
  const cols = [
    { header: 'Timestamp', key: 'timestamp', width: 20 },
    { header: 'Akcja', key: 'akcja', width: 25 },
    { header: 'Użytkownik', key: 'username', width: 18 },
    { header: 'Rola', key: 'rola', width: 12 },
    { header: 'Tabela', key: 'tabela_docelowa', width: 18 },
    { header: 'Rekord ID', key: 'rekord_id', width: 12 },
    { header: 'Jednostka', key: 'tenant_nazwa', width: 20 },
    { header: 'IP', key: 'ip_adres', width: 15 },
  ];
  if (format === 'xlsx') return sendXlsxResponse(res, cols, rows, 'Audit_log', 'Audit Log');
  if (format === 'csv') return sendCsvResponse(res, cols, rows, 'Audit_log');
  res.json(rows);
});

module.exports = router;

File: backend-py/routes/rejestry.py
from __future__ import annotations
from typing import Optional

from fastapi import APIRouter, Depends
from sqlalchemy import text
from sqlalchemy.orm import Session

from database import get_session
from dependencies import CurrentUser, require_roles

router = APIRouter()
IT = ("IT_ADMIN", "SUPERADMIN")


@router.get("/systemy")
async def rejestr_systemow(tenant_id: Optional[int] = None, format: Optional[str] = None,
                            session: Session = Depends(get_session),
                            current_user: CurrentUser = Depends(require_roles(*IT))):
    q = """SELECT s.id, s.nazwa, t.nazwa as tenant_nazwa, s.poziom_krytycznosci,
                  s.wlasciciel, s.aktywny, s.data_dodania,
                  COUNT(DISTINCT CASE WHEN up.aktywne=1 THEN up.pracownik_id END) as aktywni_uzytkownicy,
                  MAX(pr.data_zakonczenia) as ostatni_przeglad
             FROM systemy_it s
             LEFT JOIN tenants t ON t.id=s.tenant_id
             LEFT JOIN uprawnienia up ON up.system_id=s.id
             LEFT JOIN przeglady pr ON pr.tenant_id=s.tenant_id AND pr.status='ZAKOŃCZONY'
            WHERE s.aktywny=1"""
    params: dict = {}
    if tenant_id:
        q += " AND (s.tenant_id=:tid OR s.tenant_id IS NULL)"
        params["tid"] = tenant_id
    q += " GROUP BY s.id, t.nazwa ORDER BY s.poziom_krytycznosci DESC, s.nazwa"
    rows = session.execute(text(q), params).mappings().all()
    data = [dict(r) for r in rows]

    if format in ("xlsx", "csv"):
        from services.excel_service import send_xlsx, send_csv
        cols = [
            {"header": "System", "key": "nazwa"},
            {"header": "Jednostka", "key": "tenant_nazwa"},
            {"header": "Krytyczność", "key": "poziom_krytycznosci"},
            {"header": "Właściciel", "key": "wlasciciel"},
            {"header": "Aktywni użytkownicy", "key": "aktywni_uzytkownicy"},
            {"header": "Ostatni przegląd", "key": "ostatni_przeglad"},
            {"header": "Status", "key": "aktywny"},
        ]
        if format == "xlsx":
            return send_xlsx(data, cols, "Rejestr_systemow", "Systemy IT")
        return send_csv(data, cols, "Rejestr_systemow")
    return data


@router.get("/uprzywilejowani")
async def rejestr_uprzywilejowanych(tenant_id: Optional[int] = None, format: Optional[str] = None,
                                     session: Session = Depends(get_session),
                                     current_user: CurrentUser = Depends(require_roles(*IT))):
    q = """SELECT up.id, up.data_od, up.aktywne,
                  u.imie || ' ' || u.nazwisko as pracownik_nazwa,
                  t.nazwa as tenant_nazwa,
                  s.nazwa as system_nazwa,
                  z.nazwa as zakres_nazwa,
                  w.numer
             FROM uprawnienia up
             JOIN pracownicy p ON p.id=up.pracownik_id
             JOIN uzytkownicy u ON u.id=p.uzytkownik_id
             JOIN tenants t ON t.id=up.tenant_id
             JOIN systemy_it s ON s.id=up.system_id
             JOIN zakres_uprawnien z ON z.id=up.zakres_id
             LEFT JOIN wnioski w ON w.id=up.wniosek_id
            WHERE z.uprzywilejowany=1 AND up.aktywne=1"""
    params: dict = {}
    if tenant_id:
        q += " AND up.tenant_id=:tid"
        params["tid"] = tenant_id
    q += " ORDER BY t.nazwa, u.nazwisko"
    rows = session.execute(text(q), params).mappings().all()
    data = [dict(r) for r in rows]

    if format in ("xlsx", "csv"):
        from services.excel_service import send_xlsx, send_csv
        cols = [
            {"header": "Pracownik", "key": "pracownik_nazwa"},
            {"header": "Jednostka", "key": "tenant_nazwa"},
            {"header": "System", "key": "system_nazwa"},
            {"header": "Zakres", "key": "zakres_nazwa"},
            {"header": "Data nadania", "key": "data_od"},
            {"header": "Nr wniosku", "key": "numer"},
            {"header": "Aktywne", "key": "aktywne"},
        ]
        if format == "xlsx":
            return send_xlsx(data, cols, "Uprzywilejowani", "Uprzywilejowani")
        return send_csv(data, cols, "Uprzywilejowani")
    return data


@router.get("/przeglady")
async def rejestr_przeglady(tenant_id: Optional[int] = None, format: Optional[str] = None,
                             session: Session = Depends(get_session),
                             current_user: CurrentUser = Depends(require_roles(*IT))):
    q = """SELECT pr.id, pr.typ, pr.status, pr.data_rozpoczecia, pr.data_zakonczenia,
                  t.nazwa as tenant_nazwa,
                  u.imie || ' ' || u.nazwisko as inicjujacy_nazwa,
                  COUNT(CASE WHEN pp.decyzja='POZOSTAW' THEN 1 END) as pozostaw,
                  COUNT(CASE WHEN pp.decyzja='COFNIJ' THEN 1 END) as cofnij,
                  COUNT(CASE WHEN pp.decyzja='ZMODYFIKUJ' THEN 1 END) as zmodyfikuj
             FROM przeglady pr
             JOIN tenants t ON t.id=pr.tenant_id
             JOIN uzytkownicy u ON u.id=pr.inicjujacy_id
             LEFT JOIN pozycje_przegladu pp ON pp.przeglad_id=pr.id
            WHERE 1=1"""
    params: dict = {}
    if tenant_id:
        q += " AND pr.tenant_id=:tid"
        params["tid"] = tenant_id
    q += " GROUP BY pr.id, t.nazwa, u.imie, u.nazwisko ORDER BY pr.data_rozpoczecia DESC"
    rows = session.execute(text(q), params).mappings().all()
    data = [dict(r) for r in rows]

    if format in ("xlsx", "csv"):
        from services.excel_service import send_xlsx, send_csv
        cols = [
            {"header": "Nr", "key": "id"},
            {"header": "Jednostka", "key": "tenant_nazwa"},
            {"header": "Typ", "key": "typ"},
            {"header": "Status", "key": "status"},
            {"header": "Data rozp.", "key": "data_rozpoczecia"},
            {"header": "Data zak.", "key": "data_zakonczenia"},
            {"header": "Inicjujący", "key": "inicjujacy_nazwa"},
            {"header": "Pozostaw", "key": "pozostaw"},
            {"header": "Cofnij", "key": "cofnij"},
            {"header": "Zmodyfikuj", "key": "zmodyfikuj"},
        ]
        if format == "xlsx":
            return send_xlsx(data, cols, "Rejestr_przeglady", "Przeglądy")
        return send_csv(data, cols, "Rejestr_przeglady")
    return data


@router.get("/audit")
async def rejestr_audit(tenant_id: Optional[int] = None, user_id: Optional[int] = None,
                         akcja: Optional[str] = None, tabela: Optional[str] = None,
                         from_: Optional[str] = None, to: Optional[str] = None,
                         page: int = 1, limit: int = 50, format: Optional[str] = None,
                         session: Session = Depends(get_session),
                         current_user: CurrentUser = Depends(require_roles(*IT))):
    offset = (page - 1) * limit
    q = """SELECT al.id, al.timestamp, al.akcja, al.tabela_docelowa, al.rekord_id,
                  al.username, al.rola, al.ip_adres,
                  t.nazwa as tenant_nazwa,
                  al.stare_dane, al.nowe_dane
             FROM audit_log al
             LEFT JOIN tenants t ON t.id=al.tenant_id
            WHERE 1=1"""
    params: dict = {}
    if tenant_id:
        q += " AND al.tenant_id=:tid"
        params["tid"] = tenant_id
    if user_id:
        q += " AND al.user_id=:uid"
        params["uid"] = user_id
    if akcja:
        q += " AND al.akcja LIKE :a"
        params["a"] = f"%{akcja}%"
    if tabela:
        q += " AND al.tabela_docelowa=:tab"
        params["tab"] = tabela
    if from_:
        q += " AND al.timestamp>=:fr"
        params["fr"] = from_
    if to:
        q += " AND al.timestamp<=:to"
        params["to"] = to
    q += " ORDER BY al.timestamp DESC"

    if format not in ("xlsx", "csv"):
        count_q = f"SELECT COUNT(*) as cnt FROM ({q}) sub"
        total_row = session.execute(text(count_q), params).mappings().first()
        total = total_row["cnt"] if total_row else 0
        params["lim"] = limit
        params["off"] = offset
        paged_q = q + " LIMIT :lim OFFSET :off"
        rows = session.execute(text(paged_q), params).mappings().all()
        return {"data": [dict(r) for r in rows], "total": total, "page": page, "limit": limit}

    rows = session.execute(text(q), params).mappings().all()
    data = [dict(r) for r in rows]
    from services.excel_service import send_xlsx, send_csv
    cols = [
        {"header": "Timestamp", "key": "timestamp"},
        {"header": "Akcja", "key": "akcja"},
        {"header": "Użytkownik", "key": "username"},
        {"header": "Rola", "key": "rola"},
        {"header": "Tabela", "key": "tabela_docelowa"},
        {"header": "Rekord ID", "key": "rekord_id"},
        {"header": "Jednostka", "key": "tenant_nazwa"},
        {"header": "IP", "key": "ip_adres"},
    ]
    if format == "xlsx":
        return send_xlsx(data, cols, "Audit_log", "Audit Log")
    return send_csv(data, cols, "Audit_log")
